src/hex/shape.rs

```diff
@@ -51,16 +73,176 @@ impl Shape<Axial> {
+            }
+            vec
+        }
+    }
+
+    // Creates a shape.
+    // This algorithm will produce regular and irregular shapes with points from points argument.
+    // The algorithm *WILL* calculate it's inequalities on EVERY point in the array. So, in example, if you have a point
+    // inside a shape, that point will still be calculated but will not change anything about the resultant inequality.
+    // points: Coordinates to calculate the system of linear inequalities from.
+    pub fn make_shape<T>(points: &[Axial], origin: Option<Axial>, data: Option<T>, constuctor: &dyn Fn(Axial, Option<T>) -> HexTile<T>) -> HexShape<T>
+    where
+        T: Clone + Default
+    {
+        if points.is_empty() {
+            return HexShape::new(None, None);
+        }
+
+        let transform = transform!(origin.unwrap_or(points[0]));
+
+        let hexes = Inequality::new(points).solve::<HexTile<T>, T>(data, constuctor);
+
+        HexShape::new(Some(hexes), Some(transform))
+    }
+
+    pub struct ShapeArgs {
+        pub origin: Axial,
+        pub size: u32,
+        pub rot_dir: i32,
+    }
+
+    #[macro_export]
+    macro_rules! shapeargs {
+        ($o:expr, $s:expr, $rd:expr) => {
+            ShapeArgs {
+                origin: $o,
+                size: $s,
+                rot_dir: $rd,
+            }
+        };
+    }
+    pub use shapeargs;
+
+    // Creates a line.
+    // The shape has hexes originating from (0,0) in local space.
+    // origin: Denotes the grid space location the shape occupies. Affects the transform.
+    pub fn line<T>(args: ShapeArgs, data: Option<T>, constuctor: &dyn Fn(Axial, Option<T>) -> HexTile<T>) -> HexShape<T> 
+    where
+        T: Clone + Default
+    {
+        // Working in local space
+        let vertex_a = axial!(0, 0);
+        let vertex_b = vertex_a.make_vector(args.size as i32, args.rot_dir);
+        make_shape(&[vertex_a, vertex_b], Some(args.origin), data, constuctor)
+    }
+
+    // Creates a regular triangle.
+    // The shape has hexes originating from (0,0) in local space.
+    // origin: Denotes the grid space location the shape occupies. Affects the transform.
+    pub fn triangle<T>(args: ShapeArgs, data: Option<T>, constuctor: &dyn Fn(Axial, Option<T>) -> HexTile<T>) -> HexShape<T> 
+    where
+        T: Clone + Default
+    {
+        // Working in local space
+        let vertex_a = axial!(0, 0);
+        let vertex_b = vertex_a.make_vector(args.size as i32, args.rot_dir);
+        let vertex_c = vertex_a.make_vector(args.size as i32, args.rot_dir + 1);
+
+        make_shape(&[vertex_a, vertex_b, vertex_c], Some(args.origin), data, constuctor)
+    }
+
+    // Creates a regular rhombus.
+    // The shape has hexes originating from (0,0) in local space.
+    // origin: Denotes the grid space location the shape occupies. Affects the transform.
+    pub fn rhombus<T>(args: ShapeArgs, data: Option<T>, constuctor: &dyn Fn(Axial, Option<T>) -> HexTile<T>) -> HexShape<T>
```

Comment:
might wanna search and replace 'constuctor' into 'constructor' through this file

Reply:
That's funny, yea I did mess that up. It's correct now.